commit 5ef3a6ab173b6841e30dba80b1ff6735e114f12c
Author: Martin Ritter <martin.ritter@desy.de>
Date:   Wed Jul 24 17:03:24 2019 +0200

    Disable building tests due to conflicts with googletest 1.8.1

diff --git a/CMakeLists.txt b/CMakeLists.txt
index dbae5ac..abfe349 100644
--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -118,8 +118,6 @@ if(EXISTS "${CMAKE_SOURCE_DIR}/dist/CMakeLists.txt")
   add_subdirectory (dist)
 endif()
 
-add_subdirectory (test)
-
 
 #############################################################################
 ## headers
