--- src/EvtGenExternal/EvtPythia.cpp.original	2020-08-26 11:19:07.537135313 +0200
+++ src/EvtGenExternal/EvtPythia.cpp	2020-08-26 11:19:14.169119174 +0200
@@ -26,6 +26,7 @@
 #include "EvtGenBase/EvtParticle.hh"
 #include "EvtGenBase/EvtPatches.hh"
 #include "EvtGenBase/EvtSpinDensity.hh"
+#include "EvtGenBase/EvtDecayTable.hh"
 
 #include "EvtGenModels/EvtAbsExternalGen.hh"
 
@@ -83,7 +84,46 @@
     }
 
     if ( _pythiaEngine ) {
-        _pythiaEngine->doDecay( p );
+        //  _pythiaEngine->doDecay(p);
+
+        //NEW (TF, 2015) to avoid double counting when pythia produces a decay that is already defined in the decfile
+        int more = 0; //used for the check
+        int ntrials = 0; //counter for trials
+        const int maxtrials = 1000; //number of trials to regenerate event
+        const int maxdaughters = 25; //maximum number of first generation decay daughters
+
+        do {
+          _pythiaEngine->doDecay(p);
+
+          EvtId parentid = p->getId();
+          int ndaughters = p->getNDaug();
+          EvtId daughters[maxdaughters];
+
+          int i = 0;
+          for ( i = 0; i < ndaughters && ndaughters < maxdaughters; i++) {
+               EvtId daugId = p->getDaug(i)->getId();
+               daughters[i] = daugId;
+          }
+          
+          //exception maxdaughters
+          if(ndaughters >= maxdaughters) {
+               EvtGenReport(EVTGEN_ERROR, "EvtPythia::decay") << " naughters >= maxdaughters Will terminate execution!" << std::endl; ::abort();
+          }
+
+          EvtDecayTable* table = EvtDecayTable::getInstance();
+          int channel = table->inChannelList(parentid, ndaughters, daughters);
+
+          //EvtGenReport(EVTGEN_INFO, "EvtPythia::decay") << " ntrials=" << ntrials << ", parentid=" << parentid << ", ndaughters=" << ndaughters << ", channel=" << channel << std::endl;
+
+          more = ((channel!=-1)&&(channel!=p->getChannel()));
+          ntrials = ntrials + 1;
+        
+        } while(more && ntrials<maxtrials);
+        
+        //exception maxtrials
+        if(ntrials>=maxtrials) {
+          EvtGenReport(EVTGEN_ERROR, "EvtPythia::decay") << " ntrials>=maxtrials Will terminate execution!" << std::endl; ::abort();
+        }
     }
 
     this->fixPolarisations( p );
