From a186a501a48b62570bd2cd98284ff13204082ba0 Mon Sep 17 00:00:00 2001
From: Thomas Keck <thomas.keck2@kit.edu>
Date: Mon, 9 Jan 2017 09:22:08 +0100
Subject: [PATCH] Replaced outdated Cint library with Cling

---
 cmake/FindROOT.cmake | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/cmake/FindROOT.cmake b/cmake/FindROOT.cmake
index 773e735..595b44b 100644
--- a/cmake/FindROOT.cmake
+++ b/cmake/FindROOT.cmake
@@ -43,7 +43,7 @@ if(ROOT_CONFIG_EXECUTABLE)
       OUTPUT_STRIP_TRAILING_WHITESPACE)
   set(ROOT_LIBRARY_DIRS ${ROOT_LIBRARY_DIR})
 
-  set(rootlibs Core Cint RIO Net Hist Graf Graf3d Gpad Tree Rint Postscript Matrix Physics MathCore Thread)
+  set(rootlibs Core Cling RIO Net Hist Graf Graf3d Gpad Tree Rint Postscript Matrix Physics MathCore Thread)
   set(ROOT_LIBRARIES)
   foreach(_cpt ${rootlibs} ${ROOT_FIND_COMPONENTS})
     find_library(ROOT_${_cpt}_LIBRARY ${_cpt} HINTS ${ROOT_LIBRARY_DIR})
