diff --git a/libgcc/unwind-dw2-fde-dip.c b/libgcc/unwind-dw2-fde-dip.c
--- a/libgcc/unwind-dw2-fde-dip.c
+++ b/libgcc/unwind-dw2-fde-dip.c
@@ -542,24 +542,6 @@
   if (ret != NULL)
     return ret;

-  /* Use DLFO_STRUCT_HAS_EH_DBASE as a proxy for the existence of a glibc-style
-     _dl_find_object function.  */
-#if defined(DLFO_STRUCT_HAS_EH_DBASE)
-  {
-    struct dl_find_object dlfo;
-    if (_dl_find_object (pc, &dlfo) == 0 && dlfo.dlfo_eh_frame != NULL)
-      return find_fde_tail ((_Unwind_Ptr) pc, dlfo.dlfo_eh_frame,
-# if DLFO_STRUCT_HAS_EH_DBASE
-			    (_Unwind_Ptr) dlfo.dlfo_eh_dbase,
-# else
-			    0,
-# endif
-			    bases);
-    else
-      return NULL;
-    }
-#endif /* DLFO_STRUCT_HAS_EH_DBASE */
-
   data.pc = (_Unwind_Ptr) pc;
 #if NEED_DBASE_MEMBER
   data.dbase = NULL;
