--- src/geant4/source/processes/hadronic/models/cascade/cascade/src/G4CascadeInterface.cc	2014-11-20 11:15:02.000000000 +0300
+++ geant4.10.04/source/processes/hadronic/models/cascade/cascade/src/G4CascadeInterface.cc	2017-12-07 17:31:23.000000000 +0300
@@ -319,7 +326,8 @@
   }
 
   // Rotate event to put Z axis along original projectile direction
-  output->rotateEvent(bulletInLabFrame);
+  // Removed by DHW to fix bug #1990
+  // output->rotateEvent(bulletInLabFrame);
 
   copyOutputToHadronicResult();
 
@@ -457,12 +465,14 @@
 
   // Code momentum and energy -- Bertini wants z-axis and GeV units
   G4LorentzVector projectileMomentum = aTrack.Get4Momentum()/GeV;
-  
+
   // Rotation/boost to get from z-axis back to original frame
-  bulletInLabFrame = G4LorentzRotation::IDENTITY;	// Initialize
-  bulletInLabFrame.rotateZ(-projectileMomentum.phi());
-  bulletInLabFrame.rotateY(-projectileMomentum.theta());
-  bulletInLabFrame.invert();
+  // According to bug report #1990 this rotation is unnecessary and causes
+  // irreproducibility.  Verifed and fixed by DHW 27 Nov 2017
+  // bulletInLabFrame = G4LorentzRotation::IDENTITY;	// Initialize
+  // bulletInLabFrame.rotateZ(-projectileMomentum.phi());
+  // bulletInLabFrame.rotateY(-projectileMomentum.theta());
+  // bulletInLabFrame.invert();
 
   G4LorentzVector momentumBullet(0., 0., projectileMomentum.rho(),
 				 projectileMomentum.e());
