diff --git a/source/materials/src/G4SurfaceProperty.cc b/source/materials/src/G4SurfaceProperty.cc
index d501c19..608b736 100644
--- a/source/materials/src/G4SurfaceProperty.cc
+++ b/source/materials/src/G4SurfaceProperty.cc
@@ -89,7 +89,9 @@ void G4SurfaceProperty::DumpTableInfo()
 
 void G4SurfaceProperty::CleanSurfacePropertyTable()
 {
+#ifdef G4VERBOSE
   DumpTableInfo();
+#endif
   G4SurfacePropertyTable::iterator pos;
   for(pos = theSurfacePropertyTable.begin();
       pos != theSurfacePropertyTable.end(); pos++)
@@ -101,4 +103,6 @@ void G4SurfaceProperty::CleanSurfacePropertyTable()
   }
   theSurfacePropertyTable.clear();
+#ifdef G4VERBOSE
   DumpTableInfo();
+#endif
 }
