Index src/eventRecordInterfaces/PhotosHEPEVTEvent.cxx
--- src/eventRecordInterfaces/PhotosHEPEVTEvent.cxx
+++ src/eventRecordInterfaces/PhotosHEPEVTEvent.cxx
@@ -1,5 +1,6 @@
 #include "PhotosHEPEVTEvent.h"
 #include "Log.h"
+#include <cstdio>
 
 namespace Photospp
 {
Index src/eventRecordInterfaces/PhotosHEPEVTParticle.cxx
--- src/eventRecordInterfaces/PhotosHEPEVTParticle.cxx
+++ src/eventRecordInterfaces/PhotosHEPEVTParticle.cxx
@@ -1,5 +1,6 @@
 #include "PhotosHEPEVTParticle.h"
 #include "Log.h"
+#include <cstdio>
 
 namespace Photospp
 {
