diff --git a/core/unix/src/TUnixSystem.cxx b/core/unix/src/TUnixSystem.cxx
index d588f12..225b5ad 100644
--- a/core/unix/src/TUnixSystem.cxx
+++ b/core/unix/src/TUnixSystem.cxx
@@ -3570,6 +3570,9 @@ void TUnixSystem::DispatchSignals(ESignals sig)
       break;
    case kSigSystem:
    case kSigPipe:
+      //printing is not likely to work  here
+      if (sig == kSigPipe)
+         UnixResetSignal(sig);
       Break("TUnixSystem::DispatchSignals", "%s", UnixSigname(sig));
       break;
    case kSigWindowChanged:
