diff --git a/bindings/pyroot/src/Executors.cxx b/bindings/pyroot/src/Executors.cxx
index 6ccbd03..6e04f3a 100644
--- a/bindings/pyroot/src/Executors.cxx
+++ b/bindings/pyroot/src/Executors.cxx
@@ -63,7 +63,7 @@ static inline rtype GILCall##tcode(                                           \
 PYROOT_IMPL_GILCALL( void,         V )
 PYROOT_IMPL_GILCALL( UChar_t,      B )
 PYROOT_IMPL_GILCALL( Char_t,       C )
-//PYROOT_IMPL_GILCALL( Short_t,      H )
+PYROOT_IMPL_GILCALL( Short_t,      H )
 PYROOT_IMPL_GILCALL( Int_t,        I )
 PYROOT_IMPL_GILCALL( Long_t,       L )
 PYROOT_IMPL_GILCALL( Long64_t,     LL )
@@ -167,6 +167,15 @@ PyObject* PyROOT::TIntExecutor::Execute(
 }
 
 ////////////////////////////////////////////////////////////////////////////////
+/// execute <method> with argument <self, ctxt>, construct python int return value
+
+PyObject* PyROOT::TShortExecutor::Execute(
+      Cppyy::TCppMethod_t method, Cppyy::TCppObject_t self, TCallContext* ctxt )
+{
+   return PyInt_FromLong( (Short_t)GILCallH( method, self, ctxt ) );
+}
+
+////////////////////////////////////////////////////////////////////////////////
 /// execute <method> with argument <self, ctxt>, construct python long return value
 
 PyObject* PyROOT::TLongExecutor::Execute(
@@ -750,6 +759,7 @@ namespace {
    PYROOT_EXECUTOR_FACTORY( UChar )
    PYROOT_EXECUTOR_FACTORY( UCharRef )
    PYROOT_EXECUTOR_FACTORY( UCharConstRef )
+   PYROOT_EXECUTOR_FACTORY( Short )
    PYROOT_EXECUTOR_FACTORY( ShortRef )
    PYROOT_EXECUTOR_FACTORY( UShortRef )
    PYROOT_EXECUTOR_FACTORY( Int )
@@ -804,7 +814,7 @@ namespace {
       NFp_t( "const char&",        &CreateCharConstRefExecutor        ),
       NFp_t( "const signed char&", &CreateCharConstRefExecutor        ),
       NFp_t( "const unsigned char&", &CreateUCharConstRefExecutor     ),
-      NFp_t( "short",              &CreateIntExecutor                 ),
+      NFp_t( "short",              &CreateShortExecutor               ),
       NFp_t( "short&",             &CreateShortRefExecutor            ),
       NFp_t( "unsigned short",     &CreateIntExecutor                 ),
       NFp_t( "unsigned short&",    &CreateUShortRefExecutor           ),
diff --git a/bindings/pyroot/src/Executors.h b/bindings/pyroot/src/Executors.h
index a3009a7..1d65d3a 100644
--- a/bindings/pyroot/src/Executors.h
+++ b/bindings/pyroot/src/Executors.h
@@ -34,6 +34,7 @@ namespace PyROOT {
    PYROOT_DECLARE_BASIC_EXECUTOR( CharConstRef );
    PYROOT_DECLARE_BASIC_EXECUTOR( UChar );
    PYROOT_DECLARE_BASIC_EXECUTOR( UCharConstRef );
+   PYROOT_DECLARE_BASIC_EXECUTOR( Short );
    PYROOT_DECLARE_BASIC_EXECUTOR( Int );
    PYROOT_DECLARE_BASIC_EXECUTOR( Long );
    PYROOT_DECLARE_BASIC_EXECUTOR( ULong );
