From 1aad46a542fd3d0267b7b49c3cc559664061f005 Mon Sep 17 00:00:00 2001
From: Martin Ritter <martin.ritter@lmu.de>
Date: Mon, 6 Apr 2020 08:37:18 +0000
Subject: [PATCH] Disable TCling_UnloadMarker

---
 core/metacling/src/TCling.cxx | 17 -----------------
 1 file changed, 17 deletions(-)

diff --git a/core/metacling/src/TCling.cxx b/core/metacling/src/TCling.cxx
index 9d99683..6609736 100644
--- a/core/metacling/src/TCling.cxx
+++ b/core/metacling/src/TCling.cxx
@@ -195,23 +195,6 @@ extern "C" {
 #endif
 #endif
 
-//______________________________________________________________________________
-// Infrastructure to detect and react to libCling being teared down.
-//
-namespace {
-   class TCling_UnloadMarker {
-   public:
-      ~TCling_UnloadMarker() {
-         if (ROOT::Internal::gROOTLocal) {
-            ROOT::Internal::gROOTLocal->~TROOT();
-         }
-      }
-   };
-   static TCling_UnloadMarker gTClingUnloadMarker;
-}
-
-
-
 //______________________________________________________________________________
 // These functions are helpers for debugging issues with non-LLVMDEV builds.
 //
-- 
2.25.2

