diff --git a/tmva/tmva/inc/TMVA/MethodDNN.h b/tmva/tmva/inc/TMVA/MethodDNN.h
index 806c2f3..6872605 100644
--- a/tmva/tmva/inc/TMVA/MethodDNN.h
+++ b/tmva/tmva/inc/TMVA/MethodDNN.h
@@ -60,19 +60,16 @@
 #include "TMVA/DNN/Architectures/Cuda.h"
 #endif
 
-using namespace TMVA::DNN;
-
 namespace TMVA {
 
 class MethodDNN : public MethodBase
 {
-    using Architecture_t = TReference<Double_t>;
-    using Net_t          = TNet<Architecture_t>;
-    using Matrix_t       = typename Architecture_t::Matrix_t;
+   using Architecture_t = DNN::TReference<Double_t>;
+   using Net_t          = DNN::TNet<Architecture_t>;
+   using Matrix_t       = typename Architecture_t::Matrix_t;
 
 private:
-
-   using LayoutVector_t   = std::vector<std::pair<int, EActivationFunction>>;
+   using LayoutVector_t   = std::vector<std::pair<int, DNN::EActivationFunction>>;
    using KeyValueVector_t = std::vector<std::map<TString, TString>>;
 
    struct TTrainingSettings
@@ -80,7 +77,7 @@ private:
        size_t                batchSize;
        size_t                testInterval;
        size_t                convergenceSteps;
-       ERegularization       regularization;
+       DNN::ERegularization  regularization;
        Double_t              learningRate;
        Double_t              momentum;
        Double_t              weightDecay;
@@ -95,9 +92,9 @@ private:
    // general helper functions
    void     Init();
 
-   Net_t             fNet;
-   EInitialization   fWeightInitialization;
-   EOutputFunction   fOutputFunction;
+   Net_t                fNet;
+   DNN::EInitialization fWeightInitialization;
+   DNN::EOutputFunction fOutputFunction;
 
    TString                        fLayoutString;
    TString                        fErrorStrategy;
diff --git a/tmva/tmva/src/MethodDNN.cxx b/tmva/tmva/src/MethodDNN.cxx
index 84fcca6..9189128 100644
--- a/tmva/tmva/src/MethodDNN.cxx
+++ b/tmva/tmva/src/MethodDNN.cxx
@@ -62,13 +62,9 @@
 
 ClassImp(TMVA::MethodDNN)
 
-using TMVA::DNN::EActivationFunction;
-using TMVA::DNN::ELossFunction;
-using TMVA::DNN::EInitialization;
-using TMVA::DNN::EOutputFunction;
-
 namespace TMVA
 {
+   using namespace DNN;
 
 //______________________________________________________________________________
 TMVA::MethodDNN::MethodDNN(const TString& jobName,
