diff --git a/graf3d/gl/src/TGLLogicalShape.cxx b/graf3d/gl/src/TGLLogicalShape.cxx
index 0eaa1fd..e87eea5 100644
--- a/graf3d/gl/src/TGLLogicalShape.cxx
+++ b/graf3d/gl/src/TGLLogicalShape.cxx
@@ -75,7 +75,7 @@ TGLLogicalShape::TGLLogicalShape() :
    fDLBase        (0),
    fDLSize        (1),
    fDLValid       (0),
-   fDLCache       (kTRUE),
+   fDLCache       (kFALSE),
    fRefStrong     (kFALSE),
    fOwnExtObj     (kFALSE)
 {
@@ -92,7 +92,7 @@ TGLLogicalShape::TGLLogicalShape(TObject* obj) :
    fDLBase        (0),
    fDLSize        (1),
    fDLValid       (0),
-   fDLCache       (kTRUE),
+   fDLCache       (kFALSE),
    fRefStrong     (kFALSE),
    fOwnExtObj     (kFALSE)
 {
@@ -109,7 +109,7 @@ TGLLogicalShape::TGLLogicalShape(const TBuffer3D & buffer) :
    fDLBase        (0),
    fDLSize        (1),
    fDLValid       (0),
-   fDLCache       (kTRUE),
+   fDLCache       (kFALSE),
    fRefStrong     (kFALSE),
    fOwnExtObj     (kFALSE)
 {
