diff --git a/include/root/Rtypes.h b/include/root/Rtypes.h
index 276ea6c..e8e97d2 100644
--- a/include/root/Rtypes.h
+++ b/include/root/Rtypes.h
@@ -368,13 +368,13 @@ public:
 #define R__UseDummy(name) \
    class _NAME2_(name,_c) { public: _NAME2_(name,_c)() { if (name) { } } }

-#define ClassImpUnique(name, key)
+#define ClassImpUnique(name, key) _Pragma("GCC warning \"ClassImp is not needed, please remove it\"")

 /// \deprecated
 #define ClassImp(name) ClassImpUnique(name,default)

 /// Macro for Namespace
-#define NamespaceImpUnique(name, key)
+#define NamespaceImpUnique(name, key) _Pragma("GCC warning \"NamespaceImp is not needed, please remove it\"")

 #define NamespaceImp(name) NamespaceImpUnique(name,default)

@@ -398,6 +398,6 @@ public:

 #define ClassDefT2(name,Tmpl)

-#define templateClassImpUnique(name, key)
+#define templateClassImpUnique(name, key) _Pragma("GCC warning \"templateClassImp is not needed, please remove it\"")

 #define templateClassImp(name) templateClassImpUnique(name,default)
