--- tmva/tmva/src/MethodBase.cxx	2023-06-15 01:42:05.675406000 +0900
+++ tmva/tmva/src/MethodBase.cxx	2023-06-15 01:42:41.134088899 +0900
@@ -1037,7 +1037,8 @@
       s12 += t*r;
       // print progress
       Long64_t modulo = Long64_t(nevt / 100);
+      if (modulo <= 0 ) modulo = 1;
       if (ievt % modulo == 0)
          timer.DrawProgressBar(ievt);
    }
    timer.DrawProgressBar(nevt - 1);
