Author: Martin Ritter <martin.ritter@desy.de>
Date:   Mon Sep 17 16:31:24 2018 +0200

    Fix typo in configure

diff --git a/configure b/configure
index f42a2aa..898704f 100755
--- a/configure
+++ b/configure
@@ -13638,7 +13638,7 @@ if test "${enable_session+set}" = set; then :
 fi
 
 if test x"$enable_session" = "xyes"; then
-  BUILD_CFLAGS="$BUILD_CFLAGS-DSQLITE_ENABLE_SESSION -DSQLITE_ENABLE_PREUPDATE_HOOK"
+  BUILD_CFLAGS="$BUILD_CFLAGS -DSQLITE_ENABLE_SESSION -DSQLITE_ENABLE_PREUPDATE_HOOK"
 fi
 #-----------------------------------------------------------------------
 
