From 21fa6f4a509e7d7e025a02d5ae147a887c5e1b94 Mon Sep 17 00:00:00 2001
From: Konstantin Gizdov <kgizdov@gmail.com>
Date: Mon, 14 May 2018 13:10:01 +0200
Subject: [PATCH] fix for #702

---
 src/XrdApps/XrdCpFile.cc | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/src/XrdApps/XrdCpFile.cc b/src/XrdApps/XrdCpFile.cc
index 4af1be1d1f2..f440d4934c7 100644
--- a/src/XrdApps/XrdCpFile.cc
+++ b/src/XrdApps/XrdCpFile.cc
@@ -88,7 +88,7 @@ XrdCpFile::XrdCpFile(const char *FSpec, int &badURL)
    for (i = 0; i < pTnum; i++)
        {if (!strncmp(FSpec, pTab[i].pHdr, pTab[i].pHsz))
            {Protocol = pTab[i].pVal;
-            strncpy(ProtName, pTab[i].pHdr, pTab[i].pHsz-3);
+            memcpy(ProtName, pTab[i].pHdr, pTab[i].pHsz-3);
             return;
            }
        }
