--- configure
+++ configure
@@ -130,7 +130,7 @@
     HEPMCLIBLIST="-lHepMC"
     PHOTOSLIBLIST="-lPhotosCxxInterface -lPhotosFortran"
     TAUOLALIBLIST="-lTauolaCxxInterface -lTauolaFortran"
-    PYTHIALIBLIST="-lpythia8 -llhapdfdummy"
+    PYTHIALIBLIST="-lpythia8"
 
 }
 
--- evt.pdl
+++ evt.pdl
@@ -370,7 +370,7 @@
 add  p Particle  Xi_c*+                          4324  2.6466000e+00  0.0000000e+00  0.0000000e+00     3     3  0.0000000e+00       4324
 add  p Particle  anti-Sigma_b-                  -5222  5.8078000e+00  6.5821220e-06  1.0000000e-04    -3     1  0.0000000e+00      -5222
 add  p Particle  bs_0                            5301  0.0000000e+00  0.0000000e+00  0.0000000e+00    -2     0  0.0000000e+00       5301
-add  p Particle  chi_b1                         20553  9.8927000e+00  9.8928000e+00  0.0000000e+00     0     2  0.0000000e+00      20553
+add  p Particle  chi_b1                         20553  9.8927000e+00  0.0000000e+00  0.0000000e+00     0     2  0.0000000e+00      20553
 add  p Particle  Omega_c0                        4332  2.6975000e+00  0.0000000e+00  0.0000000e+00     0     1  2.0685680e-02       4332
 add  p Particle  chi_b0(2P)                    110551  1.0232500e+01  0.0000000e+00  0.0000000e+00     0     0  0.0000000e+00     110551
 add  p Particle  anti-D0                         -421  1.8648400e+00  0.0000000e+00  0.0000000e+00     0     0  1.2294489e-01       -421
--- DECAY_2010.DEC
+++ DECAY_2010.DEC
@@ -10221,28 +10221,28 @@
 Enddecay
 
 Decay K_L0
-#0.202464226 pi+     e-      nu_e                            PHSP;  #[New mode added] #[Reconstructed PDG2011]
-#0.202464226 pi-     e+      nu_e                            PHSP;  #[New mode added] #[Reconstructed PDG2011]
-#0.135033299 pi+     mu-     nu_mu                           PHSP;  #[New mode added] #[Reconstructed PDG2011]
-#0.135033299 pi-     mu+     nu_mu                           PHSP;  #[New mode added] #[Reconstructed PDG2011]
-#0.000025738 pi0     pi+     e-      nu_e                    PHSP;  #[New mode added] #[Reconstructed PDG2011]
-#0.000025738 pi0     pi-     e+      nu_e                    PHSP;  #[New mode added] #[Reconstructed PDG2011]
-#0.000006205 pi+     e-      nu_e    e+      e-              PHSP;  #[New mode added] #[Reconstructed PDG2011]
-#0.000006205 pi-     e+      nu_e    e+      e-              PHSP;  #[New mode added] #[Reconstructed PDG2011]
-#0.194795855 pi0     pi0     pi0                             PHSP;  #[New mode added] #[Reconstructed PDG2011]
-#0.125231606 pi+     pi-     pi0                             PHSP;  #[New mode added] #[Reconstructed PDG2011]
-#0.001880711 pi+     e-      nu_e    gamma                   PHSP;  #[New mode added] #[Reconstructed PDG2011]
-#0.001880711 pi-     e+      nu_e    gamma                   PHSP;  #[New mode added] #[Reconstructed PDG2011]
-#0.000277023 pi+     mu-     nu_mu   gamma                   PHSP;  #[New mode added] #[Reconstructed PDG2011]
-#0.000277023 pi-     mu+     nu_mu   gamma                   PHSP;  #[New mode added] #[Reconstructed PDG2011]
-#0.000040995 pi+     pi-     gamma                           PHSP;  #[New mode added] #[Reconstructed PDG2011]
-#0.000001262 pi0     gamma   gamma                           PHSP;  #[New mode added] #[Reconstructed PDG2011]
-#0.000000016 pi0     gamma   e+      e-                      PHSP;  #[New mode added] #[Reconstructed PDG2011]
-#0.000545653 gamma   gamma                                   PHSP;  #[New mode added] #[Reconstructed PDG2011]
-#0.000009265 e+      e-      gamma                           PHSP;  #[New mode added] #[Reconstructed PDG2011]
-#0.000000355 mu+     mu-     gamma                           PHSP;  #[New mode added] #[Reconstructed PDG2011]
-#0.000000584 e+      e-      gamma   gamma                   PHSP;  #[New mode added] #[Reconstructed PDG2011]
-#0.000000007 mu+     mu-     gamma   gamma                   PHSP;  #[New mode added] #[Reconstructed PDG2011]
+0.202464226 pi+     e-      nu_e                            PHSP;  #[New mode added] #[Reconstructed PDG2011]
+0.202464226 pi-     e+      nu_e                            PHSP;  #[New mode added] #[Reconstructed PDG2011]
+0.135033299 pi+     mu-     nu_mu                           PHSP;  #[New mode added] #[Reconstructed PDG2011]
+0.135033299 pi-     mu+     nu_mu                           PHSP;  #[New mode added] #[Reconstructed PDG2011]
+0.000025738 pi0     pi+     e-      nu_e                    PHSP;  #[New mode added] #[Reconstructed PDG2011]
+0.000025738 pi0     pi-     e+      nu_e                    PHSP;  #[New mode added] #[Reconstructed PDG2011]
+0.000006205 pi+     e-      nu_e    e+      e-              PHSP;  #[New mode added] #[Reconstructed PDG2011]
+0.000006205 pi-     e+      nu_e    e+      e-              PHSP;  #[New mode added] #[Reconstructed PDG2011]
+0.194795855 pi0     pi0     pi0                             PHSP;  #[New mode added] #[Reconstructed PDG2011]
+0.125231606 pi+     pi-     pi0                             PHSP;  #[New mode added] #[Reconstructed PDG2011]
+0.001880711 pi+     e-      nu_e    gamma                   PHSP;  #[New mode added] #[Reconstructed PDG2011]
+0.001880711 pi-     e+      nu_e    gamma                   PHSP;  #[New mode added] #[Reconstructed PDG2011]
+0.000277023 pi+     mu-     nu_mu   gamma                   PHSP;  #[New mode added] #[Reconstructed PDG2011]
+0.000277023 pi-     mu+     nu_mu   gamma                   PHSP;  #[New mode added] #[Reconstructed PDG2011]
+0.000040995 pi+     pi-     gamma                           PHSP;  #[New mode added] #[Reconstructed PDG2011]
+0.000001262 pi0     gamma   gamma                           PHSP;  #[New mode added] #[Reconstructed PDG2011]
+0.000000016 pi0     gamma   e+      e-                      PHSP;  #[New mode added] #[Reconstructed PDG2011]
+0.000545653 gamma   gamma                                   PHSP;  #[New mode added] #[Reconstructed PDG2011]
+0.000009265 e+      e-      gamma                           PHSP;  #[New mode added] #[Reconstructed PDG2011]
+0.000000355 mu+     mu-     gamma                           PHSP;  #[New mode added] #[Reconstructed PDG2011]
+0.000000584 e+      e-      gamma   gamma                   PHSP;  #[New mode added] #[Reconstructed PDG2011]
+0.000000007 mu+     mu-     gamma   gamma                   PHSP;  #[New mode added] #[Reconstructed PDG2011]
 Enddecay
 
 End
--- src/EvtGenModels/EvtPartWave.cpp
+++ src/EvtGenModels/EvtPartWave.cpp
@@ -38,7 +38,7 @@
 #include "EvtGenBase/EvtCGCoefSingle.hh"
 #include <algorithm>
 using std::endl;
-EvtPartWave::~EvtPartWave() {}
+EvtPartWave::~EvtPartWave() { delete _evalHelAmp; } // LEP
 
 std::string EvtPartWave::getName(){
 
@@ -235,8 +235,13 @@
 				getDaug(1),
 				_HBC);
 
-}
+  delete [] _lambdaA2; // LEP
+  delete [] _lambdaB2; // LEP
+  delete [] _lambdaC2; // LEP
+  for(ib=0;ib<_nB;ib++) delete [] _HBC[ib]; // LEP
+  delete [] _HBC; // LEP
 
+}
 
 void EvtPartWave::initProbMax(){
 
--- src/EvtGenExternal/EvtPythia.cpp
+++ src/EvtGenExternal/EvtPythia.cpp
@@ -16,6 +16,7 @@
 //
 //    RYD       January 8, 1997       Module created
 //    JJB       April 2011            Modified to use new Pythia8 interface
+//    FERBER, Torben   July 2015      Modified to re-generate PYTHIA events that are already defined in the decfile
 //
 //------------------------------------------------------------------------
 
@@ -26,6 +27,8 @@
 #include "EvtGenBase/EvtPDL.hh"
 #include "EvtGenBase/EvtSpinDensity.hh"
 
+#include "EvtGenBase/EvtDecayTable.hh"
+
 #include "EvtGenExternal/EvtPythia.hh"
 
 #include "EvtGenExternal/EvtExternalGenFactory.hh"
@@ -86,9 +89,49 @@
   if (_pythiaEngine == 0) {
     _pythiaEngine = EvtExternalGenFactory::getInstance()->getGenerator(EvtExternalGenFactory::PythiaGenId);
   }
-    
+
   if (_pythiaEngine != 0) {
-    _pythiaEngine->doDecay(p);
+    //     _pythiaEngine->doDecay(p);
+
+    //NEW (TF, 2015) to avoid double counting when pythia produces a decay that is already defined in the decfile
+    int more = 0; //used for the check
+    int ntrials = 0; //counter for trials
+    const int maxtrials = 1000; //number of trials to regenerate event
+    const int maxdaughters = 25; //maximum number of first generation decay daughters
+
+    do {
+      _pythiaEngine->doDecay(p);
+
+      EvtId parentid = p->getId();
+      int ndaughters = p->getNDaug();
+      EvtId daughters[maxdaughters];
+
+      int i = 0;
+      for ( i = 0; i < ndaughters && ndaughters < maxdaughters; i++) {
+	EvtId daugId = p->getDaug(i)->getId();
+	daughters[i] = daugId;
+      }
+      
+      //exception maxdaughters
+      if(ndaughters >= maxdaughters) {
+	EvtGenReport(EVTGEN_ERROR, "EvtPythia::decay") << " naughters >= maxdaughters Will terminate execution!" << std::endl; ::abort();
+      }
+
+      EvtDecayTable* table = EvtDecayTable::getInstance();
+      int channel = table->inChannelList(parentid, ndaughters, daughters);
+
+      //EvtGenReport(EVTGEN_INFO, "EvtPythia::decay") << " ntrials=" << ntrials << ", parentid=" << parentid << ", ndaughters=" << ndaughters << ", channel=" << channel << std::endl;
+
+      more = ((channel!=-1)&&(channel!=p->getChannel()));
+      ntrials = ntrials + 1;
+    
+    } while(more && ntrials<maxtrials);
+    
+    //exception maxtrials
+    if(ntrials>=maxtrials) {
+      EvtGenReport(EVTGEN_ERROR, "EvtPythia::decay") << " ntrials>=maxtrials Will terminate execution!" << std::endl; ::abort();
+    }
+    
   }
 
   this->fixPolarisations(p);
