--- src/EvtGenModels/EvtPhotosEngine.cpp
+++ src/EvtGenModels/EvtPhotosEngine.cpp
@@ -38,6 +38,7 @@
 #include <vector>
 
 using std::endl;
+using namespace Photospp;
 
 EvtPhotosEngine::EvtPhotosEngine(std::string photonType) {
 
--- src/EvtGenModels/EvtTauolaEngine.cpp
+++ src/EvtGenModels/EvtTauolaEngine.cpp
@@ -40,6 +40,7 @@
 #include <cmath>
 
 using std::endl;
+using namespace Tauolapp;
 
 EvtTauolaEngine::EvtTauolaEngine() {
 
