--- configure
+++ configure
@@ -130,7 +130,7 @@
     HEPMCLIBLIST="-lHepMC"
     PHOTOSLIBLIST="-lPhotosCxxInterface -lPhotosFortran"
     TAUOLALIBLIST="-lTauolaCxxInterface -lTauolaFortran"
-    PYTHIALIBLIST="-lpythia8 -llhapdfdummy"
+    PYTHIALIBLIST="-lpythia8"
 
 }
 
--- DECAY_2010.DEC
+++ DECAY_2010.DEC
@@ -10221,28 +10221,28 @@
 Enddecay
 
 Decay K_L0
-#0.202464226 pi+     e-      nu_e                            PHSP;  #[New mode added] #[Reconstructed PDG2011]
-#0.202464226 pi-     e+      nu_e                            PHSP;  #[New mode added] #[Reconstructed PDG2011]
-#0.135033299 pi+     mu-     nu_mu                           PHSP;  #[New mode added] #[Reconstructed PDG2011]
-#0.135033299 pi-     mu+     nu_mu                           PHSP;  #[New mode added] #[Reconstructed PDG2011]
-#0.000025738 pi0     pi+     e-      nu_e                    PHSP;  #[New mode added] #[Reconstructed PDG2011]
-#0.000025738 pi0     pi-     e+      nu_e                    PHSP;  #[New mode added] #[Reconstructed PDG2011]
-#0.000006205 pi+     e-      nu_e    e+      e-              PHSP;  #[New mode added] #[Reconstructed PDG2011]
-#0.000006205 pi-     e+      nu_e    e+      e-              PHSP;  #[New mode added] #[Reconstructed PDG2011]
-#0.194795855 pi0     pi0     pi0                             PHSP;  #[New mode added] #[Reconstructed PDG2011]
-#0.125231606 pi+     pi-     pi0                             PHSP;  #[New mode added] #[Reconstructed PDG2011]
-#0.001880711 pi+     e-      nu_e    gamma                   PHSP;  #[New mode added] #[Reconstructed PDG2011]
-#0.001880711 pi-     e+      nu_e    gamma                   PHSP;  #[New mode added] #[Reconstructed PDG2011]
-#0.000277023 pi+     mu-     nu_mu   gamma                   PHSP;  #[New mode added] #[Reconstructed PDG2011]
-#0.000277023 pi-     mu+     nu_mu   gamma                   PHSP;  #[New mode added] #[Reconstructed PDG2011]
-#0.000040995 pi+     pi-     gamma                           PHSP;  #[New mode added] #[Reconstructed PDG2011]
-#0.000001262 pi0     gamma   gamma                           PHSP;  #[New mode added] #[Reconstructed PDG2011]
-#0.000000016 pi0     gamma   e+      e-                      PHSP;  #[New mode added] #[Reconstructed PDG2011]
-#0.000545653 gamma   gamma                                   PHSP;  #[New mode added] #[Reconstructed PDG2011]
-#0.000009265 e+      e-      gamma                           PHSP;  #[New mode added] #[Reconstructed PDG2011]
-#0.000000355 mu+     mu-     gamma                           PHSP;  #[New mode added] #[Reconstructed PDG2011]
-#0.000000584 e+      e-      gamma   gamma                   PHSP;  #[New mode added] #[Reconstructed PDG2011]
-#0.000000007 mu+     mu-     gamma   gamma                   PHSP;  #[New mode added] #[Reconstructed PDG2011]
+0.202464226 pi+     e-      nu_e                            PHSP;  #[New mode added] #[Reconstructed PDG2011]
+0.202464226 pi-     e+      nu_e                            PHSP;  #[New mode added] #[Reconstructed PDG2011]
+0.135033299 pi+     mu-     nu_mu                           PHSP;  #[New mode added] #[Reconstructed PDG2011]
+0.135033299 pi-     mu+     nu_mu                           PHSP;  #[New mode added] #[Reconstructed PDG2011]
+0.000025738 pi0     pi+     e-      nu_e                    PHSP;  #[New mode added] #[Reconstructed PDG2011]
+0.000025738 pi0     pi-     e+      nu_e                    PHSP;  #[New mode added] #[Reconstructed PDG2011]
+0.000006205 pi+     e-      nu_e    e+      e-              PHSP;  #[New mode added] #[Reconstructed PDG2011]
+0.000006205 pi-     e+      nu_e    e+      e-              PHSP;  #[New mode added] #[Reconstructed PDG2011]
+0.194795855 pi0     pi0     pi0                             PHSP;  #[New mode added] #[Reconstructed PDG2011]
+0.125231606 pi+     pi-     pi0                             PHSP;  #[New mode added] #[Reconstructed PDG2011]
+0.001880711 pi+     e-      nu_e    gamma                   PHSP;  #[New mode added] #[Reconstructed PDG2011]
+0.001880711 pi-     e+      nu_e    gamma                   PHSP;  #[New mode added] #[Reconstructed PDG2011]
+0.000277023 pi+     mu-     nu_mu   gamma                   PHSP;  #[New mode added] #[Reconstructed PDG2011]
+0.000277023 pi-     mu+     nu_mu   gamma                   PHSP;  #[New mode added] #[Reconstructed PDG2011]
+0.000040995 pi+     pi-     gamma                           PHSP;  #[New mode added] #[Reconstructed PDG2011]
+0.000001262 pi0     gamma   gamma                           PHSP;  #[New mode added] #[Reconstructed PDG2011]
+0.000000016 pi0     gamma   e+      e-                      PHSP;  #[New mode added] #[Reconstructed PDG2011]
+0.000545653 gamma   gamma                                   PHSP;  #[New mode added] #[Reconstructed PDG2011]
+0.000009265 e+      e-      gamma                           PHSP;  #[New mode added] #[Reconstructed PDG2011]
+0.000000355 mu+     mu-     gamma                           PHSP;  #[New mode added] #[Reconstructed PDG2011]
+0.000000584 e+      e-      gamma   gamma                   PHSP;  #[New mode added] #[Reconstructed PDG2011]
+0.000000007 mu+     mu-     gamma   gamma                   PHSP;  #[New mode added] #[Reconstructed PDG2011]
 Enddecay
 
 End
--- src/EvtGenModels/EvtPartWave.cpp
+++ src/EvtGenModels/EvtPartWave.cpp
@@ -38,7 +38,7 @@
 #include "EvtGenBase/EvtCGCoefSingle.hh"
 #include <algorithm>
 using std::endl;
-EvtPartWave::~EvtPartWave() {}
+EvtPartWave::~EvtPartWave() { delete _evalHelAmp; } // LEP
 
 std::string EvtPartWave::getName(){
 
@@ -235,8 +235,13 @@
 				getDaug(1),
 				_HBC);
 
-}
+  delete [] _lambdaA2; // LEP
+  delete [] _lambdaB2; // LEP
+  delete [] _lambdaC2; // LEP
+  for(ib=0;ib<_nB;ib++) delete [] _HBC[ib]; // LEP
+  delete [] _HBC; // LEP
 
+}
 
 void EvtPartWave::initProbMax(){
 
