--- source/processes/optical/src/G4OpBoundaryProcess.cc	2012-02-27 10:57:01.000000000 +0900
+++ source/processes/optical/src/G4OpBoundaryProcess.cc	2012-02-27 10:38:56.000000000 +0900
@@ -148,6 +148,7 @@
         theStatus = Undefined;
 
         aParticleChange.Initialize(aTrack);
+        aParticleChange.ProposeVelocity(aTrack.GetVelocity());
 
         G4StepPoint* pPreStepPoint  = aStep.GetPreStepPoint();
         G4StepPoint* pPostStepPoint = aStep.GetPostStepPoint();
@@ -480,7 +481,9 @@
 	aParticleChange.ProposePolarization(NewPolarization);
 
         if ( theStatus == FresnelRefraction ) {
-           G4double finalVelocity = aTrack.CalculateVelocityForOpticalPhoton();
+           G4MaterialPropertyVector* groupvel =
+           Material2->GetMaterialPropertiesTable()->GetProperty("GROUPVEL");
+           G4double finalVelocity = groupvel->Value(thePhotonMomentum);
            aParticleChange.ProposeVelocity(finalVelocity);
         }
 
