Index src/eventRecordInterfaces/PhotosHEPEVTEvent.cxx
--- src/eventRecordInterfaces/PhotosHEPEVTEvent.cxx
+++ src/eventRecordInterfaces/PhotosHEPEVTEvent.cxx
@@ -1,5 +1,6 @@
 #include "PhotosHEPEVTEvent.h"
 #include "Log.h"
+#include <cstdio>
 
 namespace Photospp
 {
Index src/eventRecordInterfaces/PhotosHEPEVTParticle.cxx
--- src/eventRecordInterfaces/PhotosHEPEVTParticle.cxx
+++ src/eventRecordInterfaces/PhotosHEPEVTParticle.cxx
@@ -1,5 +1,6 @@
 #include "PhotosHEPEVTParticle.h"
 #include "Log.h"
+#include <cstdio>
 
 namespace Photospp
 {
Index src/photosCInterfaces/PhotosEvent.cxx
--- src/photosCInterfaces/PhotosEvent.cxx
+++ src/photosCInterfaces/PhotosEvent.cxx
@@ -40,10 +40,10 @@
 
 		//check for self decays
 		vector<PhotosParticle *> daughters = p->getDaughters();
-		int i=0;
-		for(i=0;i<(int)daughters.size();i++)
-			if(daughters.at(i)->getPdgID()==p->getPdgID()) break;
-		if(i!=(int)daughters.size()) continue;
+		int j=0;
+		for(j=0;j<(int)daughters.size();j++)
+			if(daughters.at(j)->getPdgID()==p->getPdgID()) break;
+		if(j!=(int)daughters.size()) continue;
 
 		Log::Debug(2)<<"Passed particle filter"<<endl;
 		filtered.push_back(p);

