--- src/TimeShower.cc  2013-04-18 23:17:03.000000000 +0200
+++ src/TimeShower.cc  2013-08-21 16:14:53.644771935 +0200
@@ -250,7 +250,11 @@                                                                                                                                           
   }
   partonSystemsPtr->setSHat( iSys, pSum.m2Calc() );

-  // Let prepare routine do the setup.    
+  // Let prepare routine do the setup.
+  dopTlimit1        = true;
+  dopTlimit2        = true;
+  dopTdamp          = false;
+  hasWeaklyRadiated = false;
   prepare( iSys, event, true);

   // Begin evolution down in pT from hard pT scale. 


