--- include/Pythia8/Basics.h
+++ include/Pythia8/Basics.h
@@ -390,9 +390,9 @@
 
   // Print a table out of two histograms with same x axis.
   friend void table(const Hist& h1, const Hist& h2, ostream& os,
-    bool printOverUnder = false) ;
+    bool printOverUnder) ;
   friend void table(const Hist& h1, const Hist& h2, string fileName,
-    bool printOverUnder = false) ;
+    bool printOverUnder) ;
 
   // Return content of specific bin: 0 gives underflow and nBin+1 overflow.
   double getBinContent(int iBin) const;
@@ -457,9 +457,9 @@
 
 // Print a table out of two histograms with same x axis.
 void table(const Hist& h1, const Hist& h2, ostream& os = cout,
-  bool printOverUnder) ;
+  bool printOverUnder = false) ;
 void table(const Hist& h1, const Hist& h2, string fileName,
-  bool printOverUnder) ;
+  bool printOverUnder = false) ;
 
 // Operator overloading with friends
 Hist operator+(double f, const Hist& h1);
