--- a/bindings/pyroot/src/Pythonize.cxx	2015-09-16 11:57:17.325612136 +0200
+++ b/bindings/pyroot/src/Pythonize.cxx	2015-09-16 11:56:17.221103980 +0200
@@ -2392,7 +2392,11 @@
 
    if ( name.find( "iterator" ) != std::string::npos ) {
       ((PyTypeObject*)pyclass)->tp_iternext = (iternextfunc)StlIterNext;
+#if PY_VERSION_HEX >= 0x03000000
+      Utility::AddToClass( pyclass, "__next__", (PyCFunction) StlIterNext, METH_NOARGS );
+#else
       Utility::AddToClass( pyclass, "next", (PyCFunction) StlIterNext, METH_NOARGS );
+#endif
 
    // special case, if operator== is a global overload and included in the dictionary
       if ( ! HasAttrDirect( pyclass, PyStrings::gCppEq, kTRUE ) )
@@ -2442,7 +2446,11 @@
       Utility::AddToClass( pyclass, "__iter__", (PyCFunction) PyObject_SelfIter, METH_NOARGS );
 
       ((PyTypeObject*)pyclass)->tp_iternext = (iternextfunc)TIterNext;
+#if PY_VERSION_HEX >= 0x03000000
+      Utility::AddToClass( pyclass, "__next__", (PyCFunction) TIterNext, METH_NOARGS );
+#else
       Utility::AddToClass( pyclass, "next", (PyCFunction) TIterNext, METH_NOARGS );
+#endif
 
       return kTRUE;
    }
