diff --git a/arich/geometry/src/GeoARICHCreator.cc b/arich/geometry/src/GeoARICHCreator.cc
index a82b29fb29..992900222a 100644
--- a/arich/geometry/src/GeoARICHCreator.cc
+++ b/arich/geometry/src/GeoARICHCreator.cc
@@ -1060,7 +1060,7 @@ namespace Belle2 {
         //
         G4TriangularFacet* facet = new G4TriangularFacet(point_1, point_2, point_3, ABSOLUTE);
         volume_solid->AddFacet((G4VFacet*) facet);
-        delete facet;
+        // delete facet;
       }
 
       volume_solid->SetSolidClosed(true);
diff --git a/geometry/src/Materials.cc b/geometry/src/Materials.cc
index 58107e38b2..94c36d4103 100644
--- a/geometry/src/Materials.cc
+++ b/geometry/src/Materials.cc
@@ -386,15 +386,15 @@ namespace Belle2 {
       // and last but not least: get rid of all materials
       B2DEBUG(50, "Cleaning G4Materials");
       auto& materials = const_cast<G4MaterialTable&>(*G4Material::GetMaterialTable());
-      for (G4Material* mat : materials) delete mat;
+      // for (G4Material* mat : materials) delete mat;
       materials.clear();
       B2DEBUG(50, "Cleaning G4Elements");
       auto& elements = const_cast<G4ElementTable&>(*G4Element::GetElementTable());
-      for (G4Element* elm : elements) delete elm;
+      // for (G4Element* elm : elements) delete elm;
       elements.clear();
       B2DEBUG(50, "Cleaning G4Isotopes");
       auto& isotopes = const_cast<G4IsotopeTable&>(*G4Isotope::GetIsotopeTable());
-      for (G4Isotope* iso : isotopes) delete iso;
+      // for (G4Isotope* iso : isotopes) delete iso;
       isotopes.clear();
       // delete material and element builder as they keep indices to materials they created :/
       delete m_nistMaterialBuilder;
